docs/hash-map.rst

```diff
@@ -0,0 +1,578 @@
+the values from the second. If the same key appears more than once then the last
+value is taken.
+
+::
+
+    HashMap.fromList []
+    > fromList []
+
+    HashMap.fromList [(1,"uno"), (1,"one"), (2,"two"), (3,"three")]
+    > fromList [(1,"one"),(2,"two"),(3,"three")]
+
+There's another incredibly useful function for constructing a map from a list::
+
+    HashMap.fromListWith :: (a -> a -> a) -> [(k, a)] -> HashMap k a
+    HashMap.fromListWith f xs = ...
+
+:haddock_short:`/Data.HashMap.Strict#fromListWith` allows you to build a map from a
+list ``xs`` with repeated keys, where ``f`` is used to "combine" (or "choose")
+values with the same key.
+
+::
+
+    -- Build a map from a list, but only keep the largest value for each key.
+    HashMap.fromListWith max [("a", 2), ("a", 1), ("b", 2)]
+    > fromList [("a",2),("b",2)]
+
+    -- Build a histogram from a list of elements.
+    HashMap.fromListWith (+) (map (\x -> (x, 1)) ["a", "a", "b", "c", "c", "c"])
+    > fromList [("a",2),("b",1),("c",3)]
+
+    -- Build a map from a list, combining the string values for the same key.
+    HashMap.fromListWith (++) [(1, "a"), (1, "b"), (2, "x"), (2, "y")]
+    > fromList [(1,"ba"),(2,"yx")]
+
+
+
+Create a list from a map
+""""""""""""""""""""""""
+
+::
+
+    HashMap.toList :: HashMap k v -> [(k, v)]
+    HashMap.toList m = ...
+
+.. NOTE::
+   ``HashMap.toList`` is **not** the same as ``Foldable.toList``; the latter is
+   equivalent to ``elems``, although is rarely useful for maps.
+
+:haddock_short:`/Data.HashMap.Strict#toList` returns a list containing the (key,
+value) pairs in the map ``m``, the order is unspecified.
+
+::
+
+    HashMap.toList (HashMap.fromList [(1,"one"), (2,"two"), (3,"three")])
+    > [(1,"one"),(2,"two"),(3,"three")]
+
+    HashMap.toList (HashMap.fromList [(1,"one"), (2,"two"), (-3,"negative three")])
+    > [(1,"one"),(2,"two"),(-3,"negative three")]
+
+
+Querying
+^^^^^^^^
+
+Lookup an entry in the map (lookup)
+"""""""""""""""""""""""""""""""""""
+
+::
+
+    HashMap.lookup :: k -> HashMap k v -> Maybe v
+    HashMap.lookup key m = ...
+
+:haddock_short:`/Data.HashMap.Strict#lookup` the value corresponding to the given
+``key``, returns ``Nothing`` if the key is not present.
+
+If you want to provide a default value if the key doesn't exist you can use:
+
+::
+
+    HashMap.lookupDefault :: v -> k -> HashMap k v -> v
+    HashMap.lookupDefault defaultVal key m = ...
+
+For example::
+
+    import Data.HashMap.Strict ((!?))
+
+    HashMap.lookup 1 HashMap.empty
+    > Nothing
+
+    HashMap.lookup 1 (HashMap.fromList [(1,"one"),(2,"two"),(3,"three")])
+    > Just "one"
+
+    > (HashMap.fromList [(1,"one"),(2,"two"),(3,"three")]) !? 1
+    > Just "one"
+
+    HashMap.lookupDefault "?" k HashMap.empty
+    > "?"
+
+    HashMap.lookupDefault "?" 1 (Map.fromList [(1,"one"), (2,"two"), (3,"three")])
+    > "one"
+
+.. WARNING::
+   **DO NOT** Use ``HashMap.!``. It is partial and throws a runtime error if
```

Comment:
This is a bit prescriptive, like saying never use `head`. Can we just say ```**DO NOT** use `HashMap.! if you are expecting a total function or cannot tolerate a runtime errors``` ?

Reply:
Reworded to be a bit of a mix of both. PTAL!